### DISCUSSION TITLE
Display satellite image using its local affine transformation matrix?

### DISCUSSION THREAD

**Discussion by tdewolff** [openlayers/openlayers#15811] [Q&A] [answered]
I'm still trying to wrap my head around projections in OpenLayers, so forgive my ignorance. I have a satellite image which is stored as a GeoTIFF file of about 900MB. I extract a subsampled version and serve it as a JPG (~200 kB) from my server. The GeoTIFF has a `ModelTransformation` tag (pixel space to WGS84 space) as well as a an `RPCCoefficient` tag (WGS84 space to pixel space). The ModelTransformation is a 4x4 3D affine transformation matrix (but the Z-axis is irrelevant, so it's essentially a 3x3 2D affine transformation matrix). The matrix is a local coordinate projection of the image onto WGS84, ie. it allows me to transform/project each pixel from its pixel space to longitude-latitude space of the WGS84 model. The ModelTransformation matrix is:

```
[4.05e-06 -1.38e-05 0 -79.44;
-1.84e-05 -2.81e-06 0 9.13;
0 0 1 0;
0 0 0 1]
```

Now I try to display the image correctly within OpenLayers, so I define a custom proj4 projection for the image as:

```
proj4.defs('custom:0', '+proj=affine +s11=4.05e-06 +s12=-1.38e-05 +s13=0 +xoff=-79.44 +s21=-1.84e-05 +s22=-2.81e-06 +s23=0 +yoff=9.13 +s31=0 +s32=0 +s33=1 +zoff=0');
const projection = ol.proj.get('custom:0');
```

But this is throwing an exception `custom:0`. Apparently, OpenLayers doesn't like the projection, but neither OpenLayers nor proj4 throws a meaningful error. What's wrong?

Of my mathematical understanding, a projection is always from one view to another, so I'm assuming that the projection I define is from the source's space to the target space, which is EPSG:3857 currently. Or to what target space should the proj4 projection be?

Some code context:
```js
import * as proj4 from '../../node_modules/proj4/dist/proj4';
import {Map, View} from '../../node_modules/ol';
import TileLayer from '../../node_modules/ol/layer/Tile';
import ImageLayer from '../../node_modules/ol/layer/Image';
import OSM from '../../node_modules/ol/source/OSM';
import ImageStatic from '../../node_modules/ol/source/ImageStatic';
import Projection from '../../node_modules/ol/proj/Projection';
import * as proj from '../../node_modules/ol/proj';
import {register as proj4Register} from '../../node_modules/ol/proj/proj4';

const map = new Map({
  target: 'map',
  view: new View({
    center: proj.fromLonLat([-79.516670, 8.983333]),
    projection: 'EPSG:3857',
    zoom: 10,
  }),
  layers: [
    new TileLayer({
      source: new OSM(),
    }),
  ],
});

const xhr = new XMLHttpRequest();
xhr.open('GET', '/img', true);
xhr.responseType = 'blob';
xhr.onload = () => {
  const proj4Def = xhr.getResponseHeader("X-Proj4");

  proj4.defs('custom:0', proj4Def);
  proj4Register(proj4);
  const projection = proj.get('custom:0');

  const urlData = URL.createObjectURL(xhr.response);
  map.addLayer(new ImageLayer({
    source: new ImageStatic({
      url: urlData,
      projection: projection,
      imageExtent: [-79.79776156987593, 8.462144626314508, -79.31111668378968, 9.138435804308466],
    }),
  }));
};
xhr.onerror = (e) => {
  console.error('XHR response from ' + e.target.responseURL);
};
xhr.send();
```

**Comment by ahocevar** [comment, marked as answer]
proj4 does not support affine transforms, but you don't need proj4 here. You can easily define and register a transform function in OpenLayers with your affine transform. This example shows how to do that (not an affine transform though, but I think you'll get the picture): https://openlayers.org/en/latest/examples/wms-custom-proj.html

**Comment by tdewolff** [comment]
Thanks @ahocevar for the quick answer! I was confused since proj.4 mentions https://proj.org/en/9.3/operations/transformations/affine.html, but that is a transformation and not a projection? Not sure if that is a difference.

Anyways, following the example I finally get the image displayed and transformed! Thank you! (final solution is below)

One question, how would I go about using RPC coefficients (a 3rd order polynomial to accurately transform from WGS84 space to pixel space) to transform the image? This is not an affine transformation, and it looks like `addCoordinateTransforms` simply transforms the image corners and thus transforms only using affine transformations. Is it at all possible to project-on-the-fly using a non-affine transformation, specifically using RPC coefficients? Or would I need to preprocess the transformation on the server before sending to JS?

```js
    const projection = new Projection({
      code: 'custom:0',
      extent: [0, 0, 4095, 3188],
      units: 'pixels',
    });
    proj.addProjection(projection);
    proj.addCoordinateTransforms(projection, 'EPSG:3857', c => {
      // TODO (not used apparently)
    }, c => {
      c = epsg3857To4326(c);
      c = transform(modelInv, c);
      return c;
    });

    map.addLayer(new ImageLayer({
      source: new ImageStatic({
        url: urlData,
        projection: projection,
        imageExtent: [0, 0, 4095, 3188],
      }),
    }));  
```

**Comment by ahocevar** [comment]
>  I was confused since proj.4 mentions https://proj.org/en/9.3/operations/transformations/affine.html

Note that proj4 is not on par with C Proj regarding features, so the correct documentation to look at is http://proj4js.org/, which does not document all the features, but links to the sources.

> it looks like addCoordinateTransforms simply transforms the image corners and thus transforms only using affine transformations.

Not at all! OpenLayers raster reprojection can do accurate transforms by using triangles. See e.g. https://openlayers.org/en/latest/examples/reprojection.html and klick "render reprojection edges".

> Is it at all possible to project-on-the-fly using a non-affine transformation, specifically using RPC coefficients?

Definitely!

> Or would I need to preprocess the transformation on the server before sending to JS?

In my opinion, it is always better to provide images on the server in the projection they will be consumed in by the client, but if you don't own the service, OpenLayers raster reprojection is a an outstanding feature.

**Reply by tdewolff** [reply]
Thanks for your kind help!
